### DISCUSSION TITLE
Lyrics

### DISCUSSION THREAD

**Discussion by josequeiros** [SubmarinerApp/Submariner#172] [Ideas]
Are there any plans on including Lyrics display from ID3 tags? Both LRC and plain text ones?

**Comment by NattyNarwhal** [comment]
When I get around to it, likely relying on the server to implement it via [`getLyrics`](http://www.subsonic.org/pages/api.jsp#getLyrics). Subsonic implements it at least, other servers may be spottier (it depends on the mood of i.e. Gracenote).

Caching lyrics/using files is an interesting idea though.

**Comment by josequeiros** [comment]
I made a POC, using navidrome.
I got LRC (timestamped lyrics) for a song, and inserted the lyrics inside
ID3 lyrics metatag inside the music file (flac in this case, but could be
mp3). I went a bit further and translated the lyrics just for
experimentation purposes.
Video attached.

[image: image.png]

**Reply by NattyNarwhal** [reply]
I don't see any video attached; can you try reuploading it? I'm assuming the mail gateway might have eaten it.

**Comment by josequeiros** [comment]
email reply did not deliver the image nor attachement: 
https://github.com/SubmarinerApp/Submariner/assets/3472128/e4027e14-8e9e-460b-8f89-a47d7c0fea1f

![image](https://github.com/SubmarinerApp/Submariner/assets/3472128/feaf088b-933c-4bb4-bbae-645486d1f39c)

**Reply by josequeiros** [reply]
@NattyNarwhal here they are (tag screenshot and video demonstration)

**Comment by NattyNarwhal** [comment]
Making notes for whenever I get around to this in the future:

* probably add lyrics attribute to track entity
  * this will probably mean switching to taglib for local file imports
* if there are no lyrics in the track, request from server
  * the subsonic API for it does look kinda clumsy (based on submitting track metadata yourself),  but most tracks are unlikely to have lyrics in metadata themselves
  * maybe we can cache this
* add a lyrics sidebar, Cmd+Y shortcut (fits with Cmd+T/U/I)
  * sidebar will just show lyrics if there is no timing information
  * if there is time-synced lyrics (how do we determine? try to parse?) do something like what apple music does, show at right time, click to jump to that time

----

A request like: `/rest/getLyrics.view?c=submariner&s=ede9e711e10637d0a69c56dd54abb3da9b5f148096e936cc58237faabd58ccd3672071e348637198497b33dfe1a8f6d3abc4d279c4f3b31081ca1badbc993db2&t=7c6c5ea4a70adc8a87a20b705d0c13ba&u=guest1&v=1.15.0&artist=MGMT&title=Kids` results in:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<subsonic-response xmlns="http://subsonic.org/restapi" status="ok" version="1.16.1">
   <lyrics artist="MGMT" title="Kids">You were a child
Crawling on your knees toward him
Making momma so proud
But your voice is too loud
We like to watch you laughing
Picking insects off of plants
No time to think of consequences

(Chorus)
Control yourself
Take only what you need from him
A family of trees wanting to be haunted
Control yourself
Take only what you need from him

The water is warm
But it's sending me shivers
A baby is born
Crying out for attention
Memories fade
Like looking through a fogged mirror
Decisions too
Decisions are made
Decisions are made and not bought
But I thought this wouldn't hurt a lot
I guess not


Chorus x6</lyrics>
</subsonic-response>
```
